Fix wheel/touch browser locking in IE and Safari

nhunzaker opened the pull request:
In order for Firefox, Edge, and Safari to optimize scroll rendering, scroll events must be attached locally.

The React synthetic event system isn't really built to handle this, attaching events locally causes a dispatch to the React Synthetic event system, which is problematic for nested scroll event handlers. To work around that, this PR adds a deduping mechanism that track what native events have already dispatched, bailing early if they have already been processed

### Testing

I have forked @nolanlawson's wheel jank example and built it with this branch's version of React. You can view the difference by doing the following:

1. Visit http://nh-react-scroll-fix.surge.sh in IE Edge or Safari
2. Using a wheel, or two-finger scroll, scroll outer container of the page for ~10 seconds.
3. Your scrolling should never be interrupted (though in Safari you _will_ see the scroll bar lock up, the page still scrolls)

---

Related issues: https://github.com/facebook/react/issues/1254#issuecomment-289296440

nhunzaker commented on src/renderers/dom/shared/ReactBrowserEventEmitter.js:
```diff
@@ -151,6 +152,39 @@ var topEventMapping = {
   topWheel: 'wheel',
```
This object is the exact same as [`EventConstants`](https://github.com/facebook/react/blob/master/src/renderers/shared/shared/event/EventConstants.js), sans the values. Has there been any thought on consolidating them into `EventConstants`?

nhunzaker replied:
:+1:

nhunzaker commented on src/renderers/dom/stack/client/ReactDOMComponent.js:
```diff
@@ -948,6 +833,7 @@ ReactDOMComponent.Mixin = {
         DOMPropertyOperations.deleteValueForProperty(getNode(this), propKey);
       }
     }
+
```
I'm going to kill some of this whitespace noise.

nhunzaker commented:
Talked through how to deal with invalid nesting with @spicyj. I'm sure there's some extra work here, but I think it's in a good spot to review.

spicyj commented on src/renderers/dom/shared/ReactBrowserEventEmitter.js:
```diff
@@ -224,7 +258,7 @@ var ReactBrowserEventEmitter = Object.assign({}, ReactEventEmitterMixin, {
-  listenTo: function(registrationName, contentDocumentHandle) {
+  listenTo: function(registrationName, contentDocumentHandle, node) {
     var mountAt = contentDocumentHandle;
     var isListening = getListeningForDocument(mountAt);
```
Can we add a 'Doc' somewhere in this name for clarity?

nhunzaker replied:
Yep!

nhunzaker commented on src/renderers/dom/shared/ReactBrowserEventEmitter.js:
```diff
@@ -151,6 +152,41 @@ var topEventMapping = {
+  topTouchMove: true,
+  topTouchStart: true,
+  topTouchCancel: true,
+  topTouchEnd: true,
```
@spicyj I've added touch cancel and touch end to the list here ☝️. They'll get attached locally now.

nhunzaker commented:
Having some trouble `npm link`ing the new flat bundled builds to the test case, but I'll get an update on http://nh-react-scroll-fix.surge.sh when possible.

gaearon commented:
I tried to fix the conflict with master, let's see if CI passes.
What's the next actionable item here?

nhunzaker commented:
I'll work on getting CI to pass later today. @spicyj I've pushed up some stuff to address your feedback, and I have a PR nearly done which consolidates the event names. It just needs to be rebased with master.

Beyond that, I wonder if it is a good idea to add some DOM fixtures for media events.

gaearon commented:
@spicyj Could you take over rebasing this PR so that @nhunzaker can continue work on https://github.com/facebook/react/pull/7311 and not worry about this one?

nhunzaker commented:
Upstreamed this with my other PR that consolidated the event list.

nhunzaker commented:
Updated with master. I think this one is good for review.

nhunzaker commented:
Rebased with master. Looks like master might have a failed test:

```
ReactCompositeComponent-state › should treat assigning to this.state inside cWM as a replaceState, with a warning

    expect(received).toEqual(expected)

    Expected value to equal:
      ["render -- step: 3, extra: false", "callback -- step: 3, extra: false"]
    Received:
      ["render -- step: 2, extra: false", "callback -- step: 2, extra: false"]
```

At least that is what I see locally. I'll watch master and see if a fix comes in, then rebase this branch when it comes through.

spicyj commented on src/renderers/dom/fiber/ReactDOMFiberComponent.js:
```diff
@@ -189,54 +163,6 @@ function trapClickOnNonInteractiveElement(node: HTMLElement) {
   node.onclick = emptyFunction;
 }
 
-function trapBubbledEventsLocal(node: Element, tag: string) {
```
Sorry for not catching this earlier – I believe these need to stay here. In particular, if you do

```
<div onLoad={fn}>
<img />
</div>
```

then the div should receive the onLoad event (at least, that's what we currently do, whether or not that makes sense) even though the img itself doesn't have a listener. Listening at the div doesn't work since in the DOM, load doesn't bubble. I'm actually surprised we don't have a test for this.

Any chance you can try to add one (maybe just render that structure, then dispatch a load event to the img node) and then fix this? I think we will just have to leave the old structure unless you have ideas.

nhunzaker commented:
@spicyj I believe I've addressed your feedback:

1. I added back the local event trapping
2. I added a test that confirms that image loading bubbles to parents, even if the image does not have a load event listener. 
3. I added a test DOM fixture for scroll. 

I've published the test DOM fixture page to http://nh-dom-fixtures-scroll.surge.sh/scroll. `local` is the build for this branch.

Interestingly, Chrome appears to no longer block on either form of scroll (window or nested)! This is super cool. I would love to know if this was a recent update to Chrome. @nolanlawson Do you know much about this?

nolanlawson commented:
I haven't heard anything from Chrome about this, but I do know it varies from desktop to mobile and there are certain "interventions" in there (e.g. touchstart/touchmove listeners are now passive by default; I believe they're considering wheel but are focusing on mobile first). Maybe @rbyers would know?

Also keep in mind that this will vary between trackpad, touchscreen, mousewheel, sidebar, and keyboard scrolling ([blog post with the gory details](https://blogs.windows.com/msedgedev/2017/03/08/scrolling-on-the-web/)). Typically `wheel` should block trackpad/mousewheel and `touchstart`/`touchmove` should block touchscreen, and nothing should block sidebar in Edge (no browser has optimized keyboard yet). Also note that trackpad scrolling varies depending on manufacturer, but external mice with wheels are usually more consistent for testing.

RByers commented:
/cc @dtapuska who leads the chromium input team.

> but I do know it varies from desktop to mobile

Nope our eventing behavior is identical across desktop and mobile (too much grey area in the middle really).

> are certain "interventions" in there (e.g. touchstart/touchmove listeners are now passive by default; I believe they're considering wheel but are focusing on mobile first). Maybe @RByers would know?

Yes, `touchstart` and `touchmove` on global elements (window, document, documentElement) are [passive by default](https://developers.google.com/web/updates/2017/01/scrolling-intervention).  That's all for now.  

We are definitely working on improving wheel scrolling performance though, and that will likely have some minor observable behavior changes (probably bringing Chrome closer to Firefox).  But we haven't done the compat analysis necessary to say for sure what change we want to make.

nhunzaker commented:
I _think_ everything is good here. Reassigning so the requested changes can be reviewed.

spicyj commented on src/renderers/dom/shared/ReactBrowserEventEmitter.js:
```diff
@@ -132,40 +139,44 @@ var ReactBrowserEventEmitter = Object.assign({}, ReactEventEmitterMixin, {
-        !(isListening.hasOwnProperty(dependency) && isListening[dependency])
+
+      if (localOnly.hasOwnProperty(dependency)) {
+        ReactDOMEventListener.trapBubbledEvent(
```
Can this be called more than once for a single node? It seems like it would since ensureListeningTo is called on updates. Let me know if I'm missing something.

nhunzaker replied:
I don't think so, because the old listener gets torn down when the property changes, and does not get reattached if it is the same. I'm adding some tests to verify.

nhunzaker replied:
I added two new tests to check for this in: 29a888a. Do you think this sufficiently covers it?

spicyj replied:
SimulateNative actually wouldn't cover it because that hooks in right at the top level listener of the event system, whereas I'm claiming that that listener would be called twice. Can you point me to where you're saying the old listener gets removed?

nhunzaker replied:
Ah okay. Also I'm wrong. It looks like this code went away:

https://github.com/facebook/react/blob/master/src/renderers/dom/stack/client/ReactDOMComponent.js#L916-L918

spicyj replied:
I can't remember us ever removing listeners actually. Might be misremembering.

nhunzaker replied:
Either way we can't re-attach listeners. I'll use `dispatchEvent` and see what I find.

gaearon commented:
@nhunzaker What are remaining items here?

nhunzaker commented:
I need about an hour and a half of digging in to figure out how to prevent local event listeners from attaching more than once. I just haven't gotten it yet.

nhunzaker commented:
Pushed an update that reproduces the issue we're talking about. Tests are going to fail. Working through some ideas.

sebmarkbage commented on src/renderers/dom/shared/__tests__/ReactBrowserEventEmitter-test.js:
```diff
@@ -435,4 +435,54 @@ describe('ReactBrowserEventEmitter', () => {
+      var el = container.querySelector('div');
+
+      // Force an update to cause a re-render
+      component.forceUpdate();
```
This is not enough to cause the ensureListeningTo to be called twice because it will bail out on function object equality here. https://github.com/facebook/react/blob/master/src/renderers/dom/fiber/ReactDOMFiberComponent.js#L672

You need to change the callback instance. For example by doing `onTouchMove={() => spy()}`.

nhunzaker replied:
👍. Good catch. I'll update this test.

If I remember, the old listener is definitely getting called. I need to work on that next.

sebmarkbage commented:
What happens if two listeners are nested inside of each other? E.g. `<div onTouchMove={...}><span onTouchMove={...} /></div>`

Will they fire twice? Because it'll bubble through both of them and the callback they invoke calls React's event system which will find two listeners along the path. So they both get executed for each call.

nhunzaker commented:
I think I got it! I've also update the fixture branch:

http://nh-react-scroll-fix.surge.sh/scroll

This makes a significant improvement to perceived scroll performance in:

- Safari
- Firefox
- IE Edge

Which is to say that the JS execution thread can be totally blocked, but the browser will at least move around the current paint rectangle (as far as I understand it).

**Dealing with the local listener duplicate attachment issue:**

I was able to resolve the local listener issue after realizing that listener attachment is tracked per-node. We can just track if a local listener is attached to a specific element without any real special behavior. 

**Could we avoid storing data on the DOM node with a WeakMap?**

I kind of wonder if we could store these listeners using a `WeakMap`, where the DOM element is the key, and the value is a `Set`, but that's not necessary for this fix (but it would be interesting to benchmark differences in memory usage and performance).

nhunzaker commented on packages/react-dom/src/events/ReactBrowserEventEmitter.js:
```diff
@@ -84,16 +84,28 @@ let reactTopListenersCounter = 0;
   }
-  return alreadyListeningTo[mountAt[topListenersIDKey]];
+  return alreadyListeningTo[node[topListenersIDKey]];
 }
```
This code relates to the `WeakMap` idea I mentioned earlier. Instead of storing data on the node itself, we could track it using a `WeakMap`. We'd still have to fallback to the node for older browsers, but I'm still curious about it.

Kovensky replied:
Doesn't Fiber already require a Map polyfill? Or does that not include WeakMap?

gaearon replied:
Just Map.

nhunzaker commented:
Working through the last piece of feedback:

> What happens if two listeners are nested inside of each other? E.g. <div onTouchMove={...}><span onTouchMove={...} /></div>

Indeed, they do dispatch twice :(. 

```javascript
<div onTouchEnd={() => console.log("top")}>
  <span onTouchEnd={() => console.log("bottom")}>Touch Here</span>
</div>
```

Touching the `span` here causes the log:

```
bottom
top
bottom
top
```

Working on that next...

gaearon commented:
Worth adding a regression test?

nhunzaker commented:
Definitely. Added in c8ff4ae6e

gaearon commented:
@nhunzaker I know it's been a while... what's the status on this?

nhunzaker commented:
Great question. 

I think the event listener still needs to be attached locally, but there needs to be some way to prevent capture and bubbling from firing multiple times for the same event source. Like it a child and parent both listen to scroll or touch.


I think one solution to this is to add the native event to a Set associated with each document that knows not to process an event more than one time during native dispatch. Then remove it from the Set when the event dispatch finishes.

That's as far as I got, but I never tried to implement it. I can take another go at this though. I've learned a lot about the event system since then.

nhunzaker commented:
Hmm. Or actually we could just not do synthetic bubbling/capture. That might be more practical

nhunzaker commented:
I think I have a working solution:

I've added a module that tracks if an event has already dispatched in React's synthetic event system. It uses a WeakSet when possible, which should automatically clean up when the event GCs. For browsers that do not support WeakSets, this module modifies the event with a flag. 

For nested scroll events (and any other locally attached events), this module prevents nested event handlers from dispatching to the synthetic event system as they bubble up the DOM.

**This needs to be tested in IE** (I'm not sure if you can modify events). I'm going to do that next.

Still, I think this approach could be used in other parts of React (like event tracking and controlled input wrapper state). Most browsers support WeakSet and WeakMap. We might be able to avoid attaching a lot of meta data to the DOM.

react-linter commented:
<!--
  0 failure: 
  0 warning: 
  
  2 markdown notices
  DangerID: danger-id-default;
-->



ReactDOM: size: :small_red_triangle:+0.3%, gzip: :small_red_triangle:+0.4%


  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: 026aa9c97813ee8e9886e2c9035a332f93c38720...bfee2c902da20822d8b90b223e96914b5dc08330</p>


  
## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | +0.1% | +0.1% | 649 KB | 649.88 KB | 152.36 KB | 152.55 KB | UMD_DEV
**react-dom.production.min.js** | **:small_red_triangle:+0.3%** | **:small_red_triangle:+0.4%** | **95.3 KB** | **95.54 KB** | **31.17 KB** | **31.29 KB** | **UMD_PROD**
react-dom.development.js | +0.1% | +0.1% | 645.14 KB | 646.02 KB | 151.24 KB | 151.45 KB | NODE_DEV
**react-dom.production.min.js** | **:small_red_triangle:+0.2%** | **:small_red_triangle:+0.3%** | **95.29 KB** | **95.52 KB** | **30.89 KB** | **30.99 KB** | **NODE_PROD**
ReactDOM-dev.js | +0.1% | +0.2% | 652.45 KB | 653.33 KB | 149.39 KB | 149.63 KB | FB_WWW_DEV
**ReactDOM-prod.js** | **:small_red_triangle:+0.6%** | **:small_red_triangle:+0.5%** | **287.9 KB** | **289.67 KB** | **53.34 KB** | **53.59 KB** | **FB_WWW_PROD**
react-dom.profiling.min.js | +0.2% | +0.4% | 96.49 KB | 96.72 KB | 31.29 KB | 31.4 KB | NODE_PROFILING
ReactDOM-profiling.js | +0.6% | +0.5% | 290.25 KB | 292.08 KB | 53.94 KB | 54.2 KB | FB_WWW_PROFILING

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

nhunzaker commented on packages/react-dom/src/events/ReactBrowserEventEmitter.js:
```diff
@@ -124,25 +125,33 @@ function getListeningForDocument(mountAt: any) {
         case TOP_SCROLL:
-          trapCapturedEvent(TOP_SCROLL, mountAt);
-          break;
+        case TOP_WHEEL:
```
I used to include touch events, but I've reduced the event list for this fix down to wheel and scroll to reduce the testing surface area. If this approach goes well, we could follow up with a PR that evaluates those events too.

nhunzaker commented on packages/react-dom/src/events/ReactDOMEventListener.js:
```diff
@@ -189,12 +193,15 @@ export function dispatchEvent(
     return;
   }
 
+  trackEventDispatch(nativeEvent);
```
This is the new logic: do not dispatch the same native event to React more than once. This prevents the following case:


```
<div onScroll={...}>
  <div onScroll={...}>
</div>
```

If a scroll event on the child triggers:

1. The child's internal React event handler dispatches to the synthetic event system
2. The child callback fires
3. The parent callback fires
4. The native event bubbles to the parent
5. The parent's internal React event handler dispatches to the synthetic event system
6. The parent callback fires

The expected behavior is that the child and parent callback are only triggered once, however the parent will trigger _twice_.

By tracking if an event has already dispatched, we can bail early and preserve the expected behavior

gaearon replied:
What if you intentionally `.dispatch()` the same native event twice? e.g. in testing?

nhunzaker replied:
It wouldn't dispatch the second time. I wonder if you could get around that by removing the event from the set when it hits the root element (or some other indicator that the event was "done" dispatching).

philipp-spiess replied:
Calling `event.stopPropagation()` might mean that it never hit the root. What if we store a list of all elements that the event already targeted and reset once a duplication is found?

I think the problem with that is that we would ignore the second dispatch if the same event targets another subtree 🤔

```
<div id="a" onScroll={...} />
<div id="b" onScroll={...} />

a.dispatchEvent(event);
b.dispatchEvent(event);
```

Now that I understand what's the issue here I can also wrap my head around that a bit 🙂

gaearon replied:
Also let's keep in mind that the end game is that we attach event listeners at the root rather than document. If that would also solve this problem maybe it's a better fix to focus on?

philipp-spiess replied:
Something we could try for sure. I'm a little worried that we overestimate the power of per-root event listener though. We seem to praise it as the solution for all of our event problems right now.

At some point I want to compile a list of issues that we've found to see which of those per-root listening can solve and what other problems might occur.

gaearon replied:
Well. I think in this case the original issue was https://github.com/facebook/react/issues/1254 which is called "React shouldn't bind wheel or touch events to the document". So that does sound to me as a specific instance of this more generic problem. :-)

nhunzaker replied:
> Also let's keep in mind that the end game is that we attach event listeners at the root rather than document. If that would also solve this problem maybe it's a better fix to focus on?

I gave that a shot, but I wasn't able to prevent scroll jank. Attaching listeners locally was the only way that I could prevent it. I can do some more testing just to be sure, though.

nhunzaker replied:
Also, if this is _only_ for scroll events, I wonder if we need to synthetically bubble/capture at all. For events that don't have synthetic dispatch phases (like mouse and composition events), maybe DOM bubbling/capture is simply sufficient.

gaearon replied:
>I gave that a shot, but I wasn't able to prevent scroll jank. Attaching listeners locally was the only way that I could prevent it. I can do some more testing just to be sure, though.

For which events? Did you include wheel, scroll, and touch events?

nhunzaker replied:
> Doesn't implement portal bubbling.

Ah shoot 😢. 

> For which events? Did you include wheel, scroll, and touch events?

Wheel and scroll. But I'll give it another shot.

gaearon replied:
Touch events seem essential to this.

nhunzaker replied:
Portal bubbling is curious to me. Portals can exist in a different tree than the parent DOM elements, but still accept bubbled events, like clicks?

gaearon replied:
https://reactjs.org/docs/portals.html#event-bubbling-through-portals

philipp-spiess replied:
Scroll is also a requirement for the responder event plugin so that might break  RNW if we remove it from the synthetic event system.

nhunzaker replied:
I sent out an alternative PR that attaches to the root here:
https://github.com/facebook/react/pull/13460

This might work, I wrote the idea off because of some issues with mouse input on BrowserStack (I think, anyway). This definitely needs testing, but the approach in #13460 definitely feels ideal if I can figure out issues in Safari.

nhunzaker commented:
Tested: this will work in IE9-11

nhunzaker commented:
Description updated, and I pushed an update to the http://nh-react-scroll-fix.surge.sh/scroll test fixture.

nhunzaker commented:
I added some extra code in b63563cd4 to attach a listener to the root element that removes tracked events. This prevents the problem of dispatching the same event twice. 

I'm still chewing on this a bit, it feels really mechanical, but I think it's starting to work towards something that feels good.

nhunzaker commented:
Looks like you definitely can't add root listeners. According to my notes:

```
** With root release

  | Browser | Outer Gesture/Wheel? | Outer Scrollbar? | Inner Gesture/Wheel? | Inner Scrollbar? |
  |---------+----------------------+------------------+----------------------+------------------|
  | Safari  | No                   | No               | No                   | No               |
  | Firefox | Yes                  | Yes              | Yes                  | Yes              |
  | Edge    | No                   | Yes              | No                   | Yes              |
  | Chrome  | Yes                  | No               | Yes                  | No               |

** Without root release

  | Browser | Outer Gesture/Wheel? | Outer Scrollbar? | Inner Gesture/Wheel? | Inner Scrollbar? |
  |---------+----------------------+------------------+----------------------+------------------|
  | Safari  | Yes                  | No               | No                   | No               |
  | Firefox | Yes                  | Yes              | Yes                  | Yes              |
  | Edge    | Yes                  | Yes              | No                   | Yes              |
  | Chrome  | Yes                  | No               | Yes                  | No               |

** It's possible that IE does not pick up wheels correctly through
   BrowserStack
```

With that in mind, I sought out to find a way to do this without attaching a listener to the root container. Turns out you can! **I altered the tracking code to keep track of the last event's current target**. This prevents duplication of event dispatch while still allowing manual dispatch of the same event during testing.

Now to do some more testing just to be sure...

nhunzaker commented on packages/react-dom/src/events/ReactEventDispatchTracker.js:
```diff
@@ -0,0 +1,28 @@
+  }
+
+  return event[lastTarget] !== event.currentTarget;
+}
```
I removed the WeakSet part of this when it became clear that I really needed to assign a value (and use a WeakMap). I've removed that from this module so that the intent is clearer. If there is a concern with assigning attributes to events, we can use a WeakMap to make the association, which should be supported in most browsers.

philipp-spiess commented:
@nhunzaker Awesome!

I can see that you're now using capturing for touch events. This will probably result in the same issues that we've found in https://github.com/facebook/react/pull/12919. Most notably there will be a breaking change in the order in which the event callbacks are fired.

nhunzaker commented:
@philipp-spiess moved touch events to bubbling and I added a test suite that verifies the bubble order on the react end (it's the same between capture and bubble, but using capture would definitely break existing expectations with events outside of react)

https://github.com/facebook/react/pull/9333/commits/e7884fce0b3c1f9e45ec3d8ad1b8476ca84f17a9

Are these valuable tests? It was useful for me to work through them, but I'm not sure if they give us any extra coverage.

philipp-spiess commented:
> Are these valuable tests? It was useful for me to work through them, but I'm not sure if they give us any extra coverage.

I think it makes sense. If we plan on handling touch events separately, we should at least make sure they still behave similarly.

nhunzaker commented:
Supporting bubbling on touch added an edge case where re-dispatching the same event below the last dispatch ignores the second dispatch incorrectly.  I've pushed up that failing case.

I'm about out of ideas, but inspiration might strike.
